microcosm_pubsub/result.py

```diff
@@ -42,69 +44,102 @@ class MessageHandlingResultType(Enum):
+            result = cls.from_error(message, error)
+        finally:
+            if result.result.retry:
+                message.nack(result.retry_timeout_seconds)
             else:
-                result = MessageHandlingResultType.SKIPPED
-            message.ack()
-        except IgnoreMessage as error:
-            extra = error.extra
-            result = MessageHandlingResultType.IGNORED
-            message.ack()
-        except SkipMessage as error:
-            extra = error.extra
+                message.ack()
+            return result
+
+    @classmethod
+    def from_result(cls, message: SQSMessage, success: bool):
+        if success:
+            result = MessageHandlingResultType.SUCCEEDED
+        else:
             result = MessageHandlingResultType.SKIPPED
-            message.ack()
-        except TTLExpired as error:
-            extra = error.extra
-            result = MessageHandlingResultType.EXPIRED
-            message.ack()
-        except Nack as nack:
-            result = MessageHandlingResultType.RETRIED
-            message.nack(nack.visibility_timeout_seconds)
-        except Exception as error:
-            exc_info_ = exc_info()
-            result = MessageHandlingResultType.FAILED
-            message.nack()
 
         return cls(
-            exc_info=exc_info_,
-            extra=extra,
             media_type=message.media_type,
             result=result,
         )
 
+    @classmethod
+    def from_error(cls, message: SQSMessage, error: Exception, **kwargs):
+        if isinstance(error, IgnoreMessage):
+            return cls(
+                extra=error.extra,
+                media_type=message.media_type,
+                result=MessageHandlingResultType.IGNORED,
+            )
+
+        if isinstance(error, SkipMessage):
+            return cls(
+                extra=error.extra,
+                media_type=message.media_type,
+                result=MessageHandlingResultType.SKIPPED,
+            )
+
+        if isinstance(error, TTLExpired):
+            return cls(
+                extra=error.extra,
+                media_type=message.media_type,
+                result=MessageHandlingResultType.EXPIRED,
+            )
+
+        if isinstance(error, Nack):
+            return cls(
+                media_type=message.media_type,
+                result=MessageHandlingResultType.RETRIED,
+                retry_timeout_seconds=error.visibility_timeout_seconds,
+            )
+
+        return cls(
+            exc_info=exc_info(),
+            media_type=message.media_type,
+            result=MessageHandlingResultType.FAILED,
+        )
+
     def log(self, logger, opaque):
-        opaque.update(self.extra)
+        """
+        Log this result.
 
-        message = f"Result for media type: {self.media_type} was : {self.result} "
+        """
+        entry = f"Result for media type: {self.media_type} was : {self.result} "
+
+        opaque.update(self.extra)
```

Comment:
Can we replace this line with:
`extra = dict(**self.extra, **opaque.as_dict())`?

Reply:
Sure.